thrust/type_traits/is_contiguous_iterator.h

```diff
@@ -212,6 +215,72 @@ struct is_contiguous_iterator_impl
     >
 {};
 
+// Type traits for contiguous iterators:
+template <typename Iterator>
+struct contiguous_iterator_traits
+{
+  static_assert(thrust::is_contiguous_iterator<Iterator>::value,
+                "contiguous_iterator_traits requires a contiguous iterator.");
+
+  using raw_pointer = typename thrust::detail::pointer_traits<
+    decltype(&*std::declval<Iterator>())>::raw_pointer;
+};
+
+template <typename Iterator>
+using contiguous_iterator_raw_pointer_t =
+  typename contiguous_iterator_traits<Iterator>::raw_pointer;
+
+// Converts a contiguous iterator to a raw pointer:
+template <typename Iterator>
+__host__ __device__
+contiguous_iterator_raw_pointer_t<Iterator>
+contiguous_iterator_raw_pointer_cast(Iterator it)
+{
+  static_assert(thrust::is_contiguous_iterator<Iterator>::value,
+                "contiguous_iterator_raw_pointer_cast called with "
+                "non-contiguous iterator.");
+  return thrust::raw_pointer_cast(&*it);
+}
+
+// Implementation for non-contiguous iterators -- passthrough.
+template <typename Iterator, bool IsContiguous>
+struct try_unwrap_contiguous_iterator_impl
```

Comment:
As I understand, we can get rid of the `impl` class here and just write:
```cpp
// Implementation for non-contiguous iterators -- passthrough.                  
template <
  typename Iterator, 
  bool IsContiguous = thrust::is_contiguous_iterator<Iterator>::value>
struct try_unwrap_contiguous_iterator_t                                         
{                                                                               
  using type = Iterator;                                                        
                                                                                
  static __host__ __device__ type get(Iterator it) { return it; }               
};                                                                              
                                                                                
// Implementation for contiguous iterators -- unwraps to raw pointer.           
template <typename Iterator>                                                    
struct try_unwrap_contiguous_iterator_t<Iterator, true /*is_contiguous*/>          
{                                                                               
  using type = contiguous_iterator_raw_pointer_t<Iterator>;                     
                                                                                
  static __host__ __device__ type get(Iterator it)                              
  {                                                                             
    return contiguous_iterator_raw_pointer_cast(it);                            
  }                                                                             
}; 
```

this makes the following code a bit simpler because we don't have to write `thrust::is_contiguous_iterator`.